ons-spark/spark-concepts/cache.md

```diff
@@ -0,0 +1,591 @@
+|   2018|            TN16|Unknown - Wild An...|             3.5|   2296.0|
+|   2017|             N19|                 Cat|             3.5|   2282.0|
++-------+----------------+--------------------+----------------+---------+
+```
+
+````{tabs}
+```{code-tab} py
+get_mean_cost_by_animal(rescue).show()
+```
+
+```{code-tab} r R
+
+get_mean_cost_by_animal(rescue) %>%
+    sparklyr::collect() %>%
+    print()
+
+```
+````
+
+```plaintext
++--------------------+------------------+
+|         AnimalGroup|          MeanCost|
++--------------------+------------------+
+|                Goat|            1180.0|
+|                Bull|             780.0|
+|                Fish|             780.0|
+|               Horse| 747.4350649350649|
+|Unknown - Animal ...| 709.6666666666666|
+|                 Cow| 624.1666666666666|
+|            Hedgehog|             520.0|
+|                Lamb|             520.0|
+|                Deer| 423.8829787234043|
+|Unknown - Wild An...|390.03636363636366|
++--------------------+------------------+
+```
+
+````{tabs}
+```{code-tab} py
+get_summary_cost_by_animal(rescue).show()
+```
+
+```{code-tab} r R
+
+get_summary_cost_by_animal(rescue) %>%
+    sparklyr::collect() %>%
+    print()
+
+```
+````
+
+```plaintext
++-----------+------+-----------------+------+-----+
+|AnimalGroup|   Min|             Mean|   Max|Count|
++-----------+------+-----------------+------+-----+
+|       Goat|1180.0|           1180.0|1180.0|    1|
+|       Fish| 260.0|            780.0|1300.0|    2|
+|       Bull| 780.0|            780.0| 780.0|    1|
+|      Horse| 255.0|747.4350649350649|3480.0|  154|
++-----------+------+-----------------+------+-----+
+```
+Look at the Spark UI. Below is the DAG for the latest query.
+
+![SQL DAG with cached data](../images/cache_dag_in_memory.png)
+
+This time Spark didn't read in the data from disk and apply the cleaning, we have broken the lineage. We know this because the DAG starts with a *InMemoryTableScan*, i.e. the cache. Note that this diagram consists of fewer stages than we had previously. Another sign of cached data being used is the presence of *Skipped Stages* on the Jobs page. 
+
+The important point here is that we were doing repeated computations on a DataFrame, i.e. aggregations on the cleansed data. When this happens it may well be more efficient to persist the data before the repeated calculations. We should check this for our case.
+
+There are many factors that contribute towards the time taken for each query that make benchmarking in Spark complex. When running the source notebook, sometimes the processing will be faster after the cache, but sometimes it will be slower. If we run the last three code cells above once more and look at the times we see that the processing after the cache was generally quicker. 
+
+Each function is colour coded in the image below. The first run was without caching, the second was with caching and the third was a second run with cached data. 
```

Comment:
I'd use shapes or numbers in addition to colour here for the benefit of those who are red/green colour blind